### DISCUSSION TITLE
How do I avoid keymapping clash with komorebi hotkeys

### DISCUSSION THREAD

**Discussion by rp1231** [houmain/keymapper#146] [Q&A] [answered]
I'm using a tiling window manager on windows 11 called komorebi.
The hotkeys I use for switching Window focus are:
`Alt + {H/J/K/L} = move focus {Left/Down/Up/Right}`

And I'm using these keymappings on keymapper:
```
[title="Stremio - Freedom to Stream" path="stremio.exe"]

H >> ArrowLeft
J >> ArrowDown
K >> ArrowUp
L >> ArrowRight
```

When I focus the stremio window, the Window focus keymappings stop working.
Even though the window focus keymappings are Alt + { Letter}
Any way to resolve this?

Thanks

**Comment by ristomatti** [comment]
Are the Window focus keymaps defined before or after those bindings? It needs to be before for these to not catch the single `HJKL`.

**Comment by rp1231** [comment]
The window focus bindings are defined in a separate app called whkd and not in keymapper

**Reply by ristomatti** [reply, marked as answer]
Keymapper rebinds the keys on a very low level and likely grabs them first. Try putting this on the top of the config and see if it fixes the issue:

```ini
NavKey = H | J | K | L
NavMod = AltLeft | AltRight

# Pass-through whkd keymaps
NavMod{NavKey} >> NavMod{NavKey}
```

**Reply by rp1231** [reply]
Thanks, That solved it for me!

**Reply by rp1231** [reply]
@ristomatti now I'm facing another problem,
the keyboard shortcuts like Alt+Shift+1 or Alt+Shift+Ctrl+1 defined in whkd aren't working.
Would you have any idea on how to solve this problem?
Thanks

**Reply by rp1231** [reply]
I have changed my config to:
```
NavMod = AltLeft | AltRight

# Pass-through whkd keymaps
NavMod{Any} >> NavMod{Any}
```
To accomodate for all the other shortcuts apart from the ones that used H|J|K|L

**Reply by rp1231** [reply]
I tried modifying NavMod to :
```
NavMod = AltLeft | AltRight | ShiftLeft | ShiftRight | ControlLeft | ControlRight
```
But that didn't help as well.

**Reply by ristomatti** [reply]
I think this should do it.

```ini
WhkdBinding = 1 | 2 # put whkd keys here
NavKey = H | J | K | L
NavMod = AltLeft | AltRight

# Pass-through whkd keymaps
(Shift NavMod){WhkdBinding} >> (Shift NavMod){WhkdBinding}
(Control NavMod){WhkdBinding} >> (Control NavMod){WhkdBinding} 
NavMod{NavKey} >> NavMod{NavKey}
```

**Reply by ristomatti** [reply]
I suggest reading https://github.com/houmain/keymapper?tab=readme-ov-file#functional-principle. The idea is to define the most specific keys first. 

But if you're not using `Alt`, `Control` or `Shift` with Keymapper, you might only need this:

```ini
AltLeft  >> AltLeft
AltRight >> AltRight
Control  >> Control
Shift    >> Shift
```

**Reply by ristomatti** [reply]
Now I'm curious. I just checked what whkd does. If all it does is execute commands on keypresses, why not use Keymapper to do that as well?

For instance, this snippet from the whkd homepage:

```
# focus windows with komorebi
alt + h : komorebic focus left
alt + j : komorebic focus down
alt + k : komorebic focus up
alt + l : komorebic focus right
```

Could as well be done in Keymapper

```ini
NavMod = AltLeft | AltRight

# focus windows with komorebi
NavMod{H} >> $(komorebic focus left)
NavMod{J} >> $(komorebic focus down)
NavMod{K} >> $(komorebic focus up)
NavMod{L} >> $(komorebic focus right)
```

**Reply by rp1231** [reply]
Makes sense. I'll try out the only keymapper config.
Thanks
